### PR TITLE
[new release] atds, atdpy, atdj, atdgen, atdgen-runtime, atdgen-codec-runtime and atd (2.3.2)

### PR DESCRIPTION
**Pull request by mjambon**
- Project page: <a href="https://github.com/ahrefs/atd">https://github.com/ahrefs/atd</a>

##### CHANGES:

* Fix package dependencies (ahrefs/atd#257)

### CODE REVIEW

**Comment by mjambon** [comment]
hmm, I don't see CI jobs being run like in my previous attempt https://github.com/ocaml/opam-repository/pull/20898

File: packages/atd/atd.2.3.2/opam [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  "dune" {>= "2.0"}
+  "menhir"
+  "easy-format"
+  "alcotest" {with-test}
+  "odoc" {with-doc}
+  "re" {>= "1.7.2"}
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atd/atd.2.3.2/opam:43]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atdgen-codec-runtime/atdgen-codec-runtime.2.3.2/opam [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+homepage: "https://github.com/ahrefs/atd"
+bug-reports: "https://github.com/ahrefs/atd/issues"
+depends: [
+  "dune" {>= "2.0"}
+  "ocaml" {>= "4.08"}
+  "odoc" {with-doc}
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atdgen-codec-runtime/atdgen-codec-runtime.2.3.2/opam:37]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atdgen-runtime/atdgen-runtime.2.3.2/opam [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  "ocaml" {>= "4.08"}
+  "dune" {>= "2.0"}
+  "yojson" {>= "1.7.0"}
+  "biniou" {>= "1.0.6"}
+  "camlp-streams"
+  "odoc" {with-doc}
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atdgen-runtime/atdgen-runtime.2.3.2/opam:40]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atdgen/atdgen.2.3.2/opam [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  "atdgen-runtime" {>= "2.0.0"}
+  "atdgen-codec-runtime" {with-test}
+  "biniou" {>= "1.0.6"}
+  "yojson" {>= "1.7.0"}
+  "odoc" {with-doc}
+  "re"
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atdgen/atdgen.2.3.2/opam:51]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atdj/atdj.2.3.2/opam [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+depends: [
+  "ocaml" {>= "4.08"}
+  "dune" {>= "2.0"}
+  "atd" {>= "2.0.0"}
+  "re"
+  "odoc" {with-doc}
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atdj/atdj.2.3.2/opam:51]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atdpy/atdpy.2.3.2/opam [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+depends: [
+  "ocaml" {>= "4.08"}
+  "dune" {>= "1.11"}
+  "alcotest" {with-test}
+  "atd" {>= "2.3.0"}
+  "cmdliner" {>= "1.1.0"}
+  "odoc" {with-doc}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atdpy/atdpy.2.3.2/opam:37]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/atds/atds.2.3.2/opam [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+bug-reports: "https://github.com/ahrefs/atd/issues"
+depends: [
+  "ocaml" {>= "4.08"}
+  "dune" {>= "2.0"}
+  "atd" {>= "2.0.0"}
+  "odoc" {with-doc}
+]
+dev-repo: "git+https://github.com/ahrefs/atd.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/atds/atds.2.3.2/opam:36]:
```suggestion
  ["dune" "subst"] {dev}
```

**Review by mseri — COMMENTED** [review]
The CI is having some issues right now. In the meantime I marked some changes that I could spot on the fly

File: packages/atd/atd.2.3.2/opam (hunk 1 above)
Comment (mjambon) [reply packages/atd/atd.2.3.2/opam:43]: I don't know what this is. It looks like it's generated by `dune` or `dune-release` from the data found in `dune-project`. Let me figure this out.

Comment (mjambon) [reply packages/atd/atd.2.3.2/opam:43]: https://opam.ocaml.org/doc/Manual.html#Package-variables says:

> `pinned`: whether the package is pinned

> `dev`: true if this is a development package, i.e. it was not built from a release archive

I'm fine with whatever works but I don't want to postprocess the output of dune-release if I don't have to. Do you know if it's a bug in dune-release or if I'm using it wrong?

Comment (mseri) [reply packages/atd/atd.2.3.2/opam:43]: Ths is a bug in the opam generation by dune, it is fixed i dune 2.7 or 2.8 but we can fix them here by hand before merging, so don’t worry. I should be able to do a proper review this evening, mentioning some more meaningful changes if needed

Comment (mjambon) [reply packages/atd/atd.2.3.2/opam:43]: I'm using dune 2.9.1, which creates or updates the local *.opam files when running `dune build`. These files contain `["dune" "subst"] {pinned}`, not `["dune" "subst"] {dev}`. It looks like it wasn't fixed or that the fix hasn't been released.

Comment (mseri) [reply packages/atd/atd.2.3.2/opam:43]: Yes but your dune file declares the dune Lang version 2.0. They backward compatibility goes all the way down to this bug

**Comment by mseri** [comment]
I think this may also be a bug in older versions of dune:
```
# File "atdgen/test/dune", line 219, characters 40-51:
# 219 |  (deps    (:atd test_annot_error.atd) %{bin:atdgen})
#                                               ^^^^^^^^^^^
# Error: Program atdgen not found in the tree or in PATH
#  (context: default)
```

File: packages/atdpy/atdpy.2.3.2/opam [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+]
+license: "MIT"
+homepage: "https://github.com/ahrefs/atd"
+bug-reports: "https://github.com/ahrefs/atd/issues"
+depends: [
+  "ocaml" {>= "4.08"}
+  "dune" {>= "1.11"}
+  "alcotest" {with-test}
+  "atd" {>= "2.3.0"}
+  "cmdliner" {>= "1.1.0"}
```
Comment (mseri) [inline packages/atdpy/atdpy.2.3.2/opam:33]:
```suggestion
  "atdgen" {>= "2.3.0"}
  "cmdliner" {>= "1.1.0"}
```

`atdpy` depends on `atdgen_emit` which is a private library that requires yojson and seems to be part of atdgen:
```
#=== ERROR while compiling atdpy.2.3.2 ========================================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(https://github.com/ahrefs/atd/releases/download/2.3.2/atdgen-codec-runtime-2.3.2.tbz)
# path                 ~/.opam/4.08/.opam-switch/build/atdpy.2.3.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p atdpy -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/atdpy-4616-761d80.env
# output-file          ~/.opam/log/atdpy-4616-761d80.out
### output ###
# File "atdgen/src/dune", line 5, characters 38-44:
# 5 |  (libraries atd biniou easy-format re yojson))
#                                           ^^^^^^
# Error: Library "yojson" not found.
# -> required by library "atdgen_emit" in _build/default/atdgen/src
# -> required by executable Atdpy_main in atdpy/src/bin/dune:2
# -> required by _build/default/atdpy/src/bin/Atdpy_main.exe
# -> required by _build/install/default/bin/atdpy
# -> required by _build/default/atdpy.install
# -> required by alias install
```

File: packages/atd/atd.2.3.2/opam (hunk 1 above)
Comment (mjambon) [reply packages/atd/atd.2.3.2/opam]: I just changed to dune lang version 2.8 and it fixes one occurrence of `["dune" "subst"] {pinned}` to `["dune" "subst"] {dev}` (in `atdpy.opam`) but not in the other opam files. I couldn't see why. I'll fix the other occurrences manually.

**Comment by mjambon** [comment]
I'm struggling to figure out why old versions of various packages are being installed and why we're getting errors. I'm surprised because we don't have these problems in [atd's CI](https://app.circleci.com/pipelines/github/ahrefs/atd?branch=master) which builds and tests everything for ocaml 4.08 and 4.13 on Ubuntu ([config](https://github.com/returntocorp/atd/tree/master/.circleci)).

**Comment by mjambon** [comment]
I added version constraints and it's starting to look better. We're still getting errors due to `python3` being not found. Python and some tools that are installable with `pip` would need to be installed for atdpy's tests to work. Also, we should have similar errors for atds and atdj which expect Scala and Java environments respectively. What's the best way to go about this? Some options I can think of:
* do nothing, leaving some errors in opam-repository's CI
* set up a flag of some kind to run only the tests that don't require these non-ocaml tools
* install the missing packages
* install the missing packages only in one environment if it's easier (e.g. Ubuntu)

All the tests are run by atd's own CI, so there's no strong need to run them here. It would be nice to have only green check marks, though. Thoughts?

**Comment by mseri** [comment]
> I'm struggling to figure out why old versions of various packages are being installed and why we're getting errors. I'm surprised because we don't have these problems in [atd's CI](https://app.circleci.com/pipelines/github/ahrefs/atd?branch=master) which builds and tests everything for ocaml 4.08 and 4.13 on Ubuntu ([config](https://github.com/returntocorp/atd/tree/master/.circleci)).

The CI in the opam repository has a lower bounds test which attempts to install the package with a configuration using the oldest allowed package. The first time you hit it, it takes a while, but after that you have a much more robust set of metadata, and this will not be an issue for future releases then

**Comment by mseri** [comment]
Almost all the remaining failures seem to need an upper bound on atdgen:


There is only one that is related to this release:
```
#=== ERROR while compiling atdgen.2.2.1 =======================================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.13.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.13/.opam-switch/build/atdgen.2.2.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p atdgen -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/atdgen-2601-a2703f.env
# output-file          ~/.opam/log/atdgen-2601-a2703f.out
### output ###
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -w -27 -safe-string -g -I atdgen/src/.atdgen_emit.objs/byte -I atdgen/src/.atdgen_emit.objs/native -I /home/opam/.opam/4.13/lib/atd -I /home/opam/.opam/4.13/lib/biniou -I /home/opam/.opam/4.13/lib/easy-format -I /home/opam/.opam/4.13/lib/yojson -intf-suffix .ml -no-alias-deps -open Atdgen_emit -o atdgen/src/.atdgen_emit.objs/native/atdgen_emit__Ocaml.cmx -c -impl atdgen/src/ocaml.ml)
# File "atdgen/src/ocaml.ml", line 618, characters 12-24:
# 618 | let split = Re.Str.split (Re.Str.regexp " ")
#                   ^^^^^^^^^^^^
# Error: Unbound module Re
```
This seems due to re being used as transitive dependency in atdgen 2.2.1. Not sure from where it was picking it so I cannot really suggest the correct fix

**Comment by mjambon** [comment]
`Error: Unbound module Re` is the same error message I got last week at some point when installing atdgen locally. Eventually, I couldn't reproduce the error (https://github.com/ocaml/opam-repository/pull/20906). I'll try to fix it now.

**Comment by mseri** [comment]
Yes, I remember your message. In this case I could check that it is compiling `atdgen.2.2.1` using `atd.2.3.2`: the issue comes from there, `atd.2.2.1` links against `re` and transitively provides it to `atdgen.2.2.1` (which is not specifying re in the libraries stanza in dune). The solution is to add an upper bound on atd in atdgen 2.2.1, but it is also why I was suggesting to start releasing them with strict versioning `=version`. All these issues would then disappear.

**Comment by mseri** [comment]
I am sending a PR for atdgen 2.0 and 2.2, will be there soon

**Comment by mseri** [comment]
There is a whole lot of
```
# Invalid or misplaced annotation <json ...  >
# Invalid or misplaced annotation <ocaml ... field_prefix... >
```
Is this breakage expected?

**Comment by mseri** [comment]
Once https://github.com/ocaml/opam-repository/pull/20936 is merged, we can rebase this and check if things are fine

**Comment by mjambon** [comment]
Regarding the errors like this one:
```
File "github.atd", line 6, characters 9-30:
Invalid or misplaced annotation <ocaml ... field_prefix... >
```
https://github.com/ahrefs/atd/pull/258 will fix the above because it's a bug in the new validation system. Other validation errors are legitimate bugs in atd files. It's a good thing to avoid breaking the reverse dependencies for now.

**Comment by mseri** [comment]
Thanks, I have updated the relevant packages in the other PR to be incompatible only with 2.3.0-2.3.2

**Comment by mjambon** [comment]
Rebased.

**Comment by mjambon** [comment]
I'm assuming we're there or almost there. I'm preparing bugfix release 2.3.3, which will be essentially the same.

**Comment by mseri** [comment]
Should we actually close this in favour of 2.3.3? Or you would like both to get in?

**Comment by mjambon** [comment]
Sure, let's close this one.

**Pull request closed without merging**